air/tests/test_module/instructions/ap.rs

```diff
@@ -185,3 +187,32 @@ fn ap_with_dst_stream() {
     assert_eq!(actual_trace, expected_state);
     assert!(result.next_peer_pks.is_empty());
 }
+
+#[test]
+fn par_ap_behaviour() {
+    let vm_1_peer_id = "vm_1_peer_id";
+    let vm_2_peer_id = "vm_2_peer_id";
+    let vm_3_peer_id = "vm_3_peer_id";
+    let mut vm_1 = create_avm(echo_call_service(), vm_1_peer_id);
+
+    let script = format!(
+        r#"
+        (par
+            (call "{2}" ("peer" "timeout") [] join_it)
+            (seq
+                (par
+                    (call "{0}" ("peer" "timeout") [join_it] $result)
+                    (ap "fast_result" $result)
+                )
+                (call "{1}" ("op" "return") [$result.$[0]])
+            )
+        )
+        "#,
+        vm_1_peer_id, vm_2_peer_id, vm_3_peer_id
+    );
+
+    let mut result = checked_call_vm!(vm_1, "", script, "", "");
+    let actual_next_peers: HashSet<_> = result.next_peer_pks.drain(..).collect();
+    let expected_next_peers: HashSet<_> = maplit::hashset!(vm_2_peer_id.to_string(), vm_3_peer_id.to_string());
+    assert_eq!(actual_next_peers, expected_next_peers);
```

Comment:
This isn't a complete test of this script. It only checks the first execution, but there's more. After checking next_peer_pks, you need to continue executing the particle. 

1. from vm_1 it hops to vm_2, vm_3
2. on vm_3, it sets `join_it` and hops to vm_2
3. on vm_2, it waits for `join_it`
4. on vm_2, it writes to `$result` and hops to vm_1
5. vm_1 executes `op return`

Originally, the problem happens on step 4: it doesn't hop `vm_2 => vm_1`